n3fit/src/n3fit/hyper_optimization/rewards.py

```diff
@@ -43,6 +44,33 @@
 log = logging.getLogger(__name__)
 
 
+def _average_best(fold_losses: np.ndarray, percentage: float = 0.9, axis: int = 0) -> float:
+    """
+    Compute the average of the input array along the specified axis, among the best `percentage`
+    of replicas.
+
+    Parameters
+    ----------
+        fold_losses: np.ndarray
+            Per replica losses for a single fold.
+        percentage: float
+            The percentage of best replicas to take into account (rounded up).
+        axis: int, optional
+            Axis along which the mean is computed. Default is 0.
+
+    Returns
+    -------
+        float: The average along the specified axis.
+    """
+    num_best = int(np.ceil(percentage * len(fold_losses)))
+
+    if np.isnan(fold_losses).any():
+        log.warning(f"{np.isnan(fold_losses).sum()} replicas have NaNs losses")
+    sorted_losses = np.sort(fold_losses, axis=axis)
+    best_losses = sorted_losses[:num_best]
+    return np.average(best_losses, axis=axis).item()
```

Comment:
We can do this for now, but I think ideally we should remove any replica more than X standard deviations from the average chi2 and if more than 10% are discarded then fail the test.

Reply:
Specifically, why don't we use the `validphys.fitveto.determine_vetoes` function with the same defaults as the `postfit` CLI for that? We know from nnpdf40 that only ~1% of replicas are discarded in postfit, so as an additional constraint we could decide to discard e.g. all setups that result in discarding more than 10% of replicas (or even fewer).

Reply:
> I don't understand why the threshold has to go from 2 to 1000 ? Surely this has to do with some 1/N that has not been considered? (where N might be replicas, data points, fold, or a combination of the above)

This is only in the hyperopt tests, what changes in this PR is not the value of the loss but what happens if it is violated, namely to fail the trial. The fact that it's violated is no surprise as it is run only for a small number of epochs in the test.

> Specifically, why don't we use the validphys.fitveto.determine_vetoes function with the same defaults as the postfit CLI for that?

You mean keep using the average in the loss as before, but in the passing criterion inside modeltrainer use this `determine_vetoes`? That's fine by me, not sure where to get these thresholds and the `fitinfos` from though.